compiler/optimizer/LocalOpts.cpp

```diff
@@ -7656,95 +7656,615 @@ int32_t TR_CheckcastAndProfiledGuardCoalescer::perform()
+                  if (op.getOpCodeValue() == TR::astore)
+                     {
+                     TR::Node *val = node->getChild(0);
+                     if (fresh.contains(val) && val->getSymbolReference() == dest)
+                        break; // store does nothing, ignore
+
+                     if (dest == castObjAuto)
+                        {
+                        castObjAuto = NULL; // it is no longer possible to remat
+                        if (trace())
+                           {
+                           traceMsg(
+                              comp(),
+                              "astore n%un [%p] updates #%d, "
+                              "which may now differ from n%un [%p]\n",
+                              node->getGlobalIndex(),
+                              node,
+                              dest->getReferenceNumber(),
+                              castObj->getGlobalIndex(),
+                              castObj);
+                           }
+                        }
+
+                     auto entry = freshByAuto.find(dest->getReferenceNumber());
+                     if (entry != freshByAuto.end())
+                        {
+                        NodeList &loads = entry->second;
+                        for (auto it = loads.begin(); it != loads.end(); ++it)
+                           {
+                           TR::Node *load = *it;
+                           TR_ASSERT_FATAL(
+                              fresh.contains(load),
+                              "expected n%un [%p] to be a fresh load of #%d",
+                              load->getGlobalIndex(),
+                              load,
+                              dest->getReferenceNumber());
+
+                           fresh.remove(load);
+                           }
+
+                        loads.clear();
+                        }
+                     }
+
+                  // Now we only need to determine whether to invalidate the
+                  // current candidate checkcast (if any).
+                  if (checkcastTree == NULL)
+                     break; // nothing to do
+
+                  // Stores to autos are OK as long as the stored auto is dead
+                  // after taking an exception from checkcast. If it's live on
+                  // the exception path, then the exception path will see the
+                  // new value even though it's supposed to see the old one,
+                  // since the exception is supposed to have been thrown before
+                  // the store. We could get more opportunities by computing
+                  // liveness, but for the moment we know that an auto is dead
+                  // on exception if there is no exception path.
+                  if (castCatchBlock == NULL)
+                     break; // dest is dead on exception, ok to reorder
+
+                  // dest may be live on exception, cannot reorder
```

Comment:
If an auto (or any sym ref for that matter) may be used in code reachable from some catch block, then it should be in the use-only aliases set for the sym ref that could cause an exception. This may be one way to get more precise at this spot but it is an optional enhancement.

Reply:
Is the use-only set sensitive to actual liveness? My impression was that it's very imprecise, in part because I don't think we compute liveness just in order to find the 'aliases,' but also because the possibly throwing symref will be shared by potentially many other exception points

Reply:
Oh ok, sounds like it's more precise than I thought, so that could be a useful in-between point

Reply:
Please let me know if you plan to update this PR for this. Skipping it in this PR or even altogether until it is needed is fine with me, I just wanted to ask so that I know if I should wait until launching more tests with the PR as it stands.

Reply:
Ah, taking another look I see that your suggestion should be very straightforward to implement, so I'll go ahead and update this